src/base64.cr

```diff
@@ -199,50 +188,179 @@ module Base64
+      output += 1
+    end
+
+    (output.address &- initial_output.address).to_i32!
+  end
+
+  # Internal method for encoding *pairs* times full 3-byte data pairs into 4-byte base64 pairs, without padding.
+  #
+  # *input* must have at least `pairs * 3` bytes available to process,
+  # while *output* must have at least `pairs * 4` bytes of storage available.
+  private def encode_base64_full_pairs_internal(input : UInt8*, output : UInt8*, pairs : Int32, chars : UInt8*) : Nil
+    # On most archivectures supported by crystal, unaligned memory access is very cheap.
+    # This section thus tries to improve performance by unrolling the loop and by replacing
+    # three aligned UInt8 accesses by one unaligned UInt32 access (discarding the last byte).
+    # The condition `pairs > 8` makes sure that there's at least one pair which is processed byte by byte,
+    # so we never accidentally read one byte further than we're allowed to (-> possible segfault).
+    #
+    # NOTE: On weak-memory architectures like risc-v, llvm must replace the
+    # unaligned UInt32 access by *four* aligned UInt8 accesses, degrading performance.
+    while pairs > 8
+      i = 8
+      while i != 0
+        value = 0_u32
+        Intrinsics.memcpy(pointerof(value), input, 4, is_volatile: false)
+        input += 3
+
+        value = value.byte_swap
+
+        output[0] = chars[(value >> 26) & 63]
+        output[1] = chars[(value >> 20) & 63]
+        output[2] = chars[(value >> 14) & 63]
+        output[3] = chars[(value >> 8) & 63]
+        output += 4
+        i &-= 1
       end
-    elsif pd == 2
-      n = (cstr.value.to_u32 << 16) | ((cstr + 1).value.to_u32 << 8)
-      yield bytes[(n >> 18) & 63]
-      yield bytes[(n >> 12) & 63]
-      yield bytes[(n >> 6) & 63]
-      yield PAD if pad
+      pairs &-= 8
+    end
+
+    while pairs > 0
+      value = 0_u32
+      Intrinsics.memcpy(pointerof(value), input, 3, is_volatile: false)
+      input += 3
+
+      value = value.byte_swap
+
+      output[0] = chars[(value >> 26) & 63]
+      output[1] = chars[(value >> 20) & 63]
+      output[2] = chars[(value >> 14) & 63]
+      output[3] = chars[(value >> 8) & 63]
+      output += 4
+      pairs &-= 1
+    end
+  end
+
+  # Internal method for encoding the final 1-2 bytes of an input to base64 with or without padding.
+  # Returns the amount of bytes written into output.
+  #
+  # This method assumes that `1 <= input_size <= 2`.
+  # Otherwise, the method's behaviour is undefined.
+  #
+  # If `pad == `true`, exactly 4 bytes will be written to *output*.
+  # Otherwise, `(input_size + 1)` bytes will be written to *output*.
+  private def encode_base64_final_pair_internal(input : UInt8*, input_size : Int32, output : UInt8*, chars : UInt8*, *, pad : Bool = false) : Int32
+    in0 = input[0]
+    output[0] = chars[in0 >> 2]
+
+    if input_size == 1
+      output[1] = chars[in0 << 6 >> 2]
+      return 2 unless pad
+
+      output[2] = PAD
+      output[3] = PAD
+    else
```

Comment:
this `else` only for `input_size == 2`, what if input_size != 1 or 2, I think need to check this and raise error.